riscv/src/compiler.rs

```diff
@@ -904,6 +904,27 @@ fn process_instruction(instr: &str, args: &[Argument]) -> Vec<String> {
             let (rd, r1, r2) = rrr(args);
             only_if_no_write_to_zero(format!("tmp1, {rd} <== mul({r1}, {r2});"), rd)
         }
+        "mulhsu" => {
+            let (rd, r1, r2) = rrr(args);
+            only_if_no_write_to_zero_vec(
+                vec![
+                    format!("tmp1 <== to_signed({r1});"),
+                    // tmp2 is 1 if tmp1 is non-negative
+                    "tmp2 <== is_positive(tmp1 + 1);".into(),
+                    // If negative, convert to positive
+                    "skip_if_zero 0, tmp2;".into(),
+                    "tmp1 <=X= 0 - tmp1;".into(),
+                    format!("tmp1, {rd} <== mul(tmp1, {r2});"),
+                    // If was negative before, convert back to negative
+                    "skip_if_zero (1-tmp2), 2;".into(),
+                    "tmp1 <== is_equal_zero(tmp1);".into(),
+                    // If the lower bits are zero, return the two's complement,
+                    // otherwise return one's complement.
+                    format!("{rd} <== wrap_signed(-{rd} - 1 + tmp1);"),
```

Comment:
I don't understand, doesn't this always negate `rd`?

Reply:
Yes, this always negate. Also, it either adds 0 or 1 to the negation (the value of `tmp1), because that is the "carry" from the two's complement of the lower bits.

But depending whether the value was originally negative, this instruction and the previous are skipped entirely by the `skip_if_zero`.

Reply:
So `skip_if_zero 1` acutally skips 2 instructions?

Reply:
oh sorry, the second argument is the number of instructions to skip!